juju/juju
# feat(machine): implement keep instance on new machine domain

This patch adds the functionality of keep-instance to the new machine domain.

Since the two methods `SetKeepInstance()` and `KeepInstance()` were only used in two facades, the call to the legacy state has been already deleted and therefore the entire functionality lives in Dqlite without the need for double writing.

_Note: Since this feature is so simple and self-contained, the whole functionality (from ddl changes to facades wiring) has been implemented in this only patch, without the need for any temporary double writes._

## Checklist


## QA steps

The keep-instance functionality must continue working as usual, no change expected even though the full functionality lives in dqlite now.

Example on lxd:

```
juju bootstrap localhost c
juju add-model m
juju add-machine
lxc ls
+---------------+---------+-----------------------+------+-----------+-----------+
|     NAME      |  STATE  |         IPV4          | IPV6 |   TYPE    | SNAPSHOTS |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-365cdd-0 | RUNNING | 10.170.236.168 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-616657-0 | RUNNING | 10.170.236.132 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+

juju remove-machine 0 --keep-instance
WARNING This command will perform the following actions:
will remove machine 0 (but retaining cloud instance)

Continue [y/N]? y
lxc ls
+---------------+---------+-----------------------+------+-----------+-----------+
|     NAME      |  STATE  |         IPV4          | IPV6 |   TYPE    | SNAPSHOTS |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-365cdd-0 | RUNNING | 10.170.236.168 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-616657-0 | RUNNING | 10.170.236.132 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
```
Then we add another machine and when we delete it without --keep-instance, we expect it to be removed from lxd.
```
juju add-machine
lxc ls
+---------------+---------+-----------------------+------+-----------+-----------+
|     NAME      |  STATE  |         IPV4          | IPV6 |   TYPE    | SNAPSHOTS |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-365cdd-0 | RUNNING | 10.170.236.168 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-616657-0 | RUNNING | 10.170.236.132 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-616657-1 | RUNNING | 10.170.236.27 (eth0)  |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
juju remove-machine 1
WARNING This command will perform the following actions:
will remove machine 1

Continue [y/N]? y
lxc ls
+---------------+---------+-----------------------+------+-----------+-----------+
|     NAME      |  STATE  |         IPV4          | IPV6 |   TYPE    | SNAPSHOTS |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-365cdd-0 | RUNNING | 10.170.236.168 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
| juju-616657-0 | RUNNING | 10.170.236.132 (eth0) |      | CONTAINER | 0         |
+---------------+---------+-----------------------+------+-----------+-----------+
```

## Links


**Jira card:** JUJU-6702
